sql: TestSessionMigration failed

sql.TestSessionMigration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20708377?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20708377?buildTab=artifacts#/) on master @ [061d9a3b2ad44d8cf7984be70cb713390df775e2](https://github.com/cockroachdb/cockroach/commits/061d9a3b2ad44d8cf7984be70cb713390df775e2):


```
=== RUN   TestSessionMigration
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/17b9b212a9f8f930df1697330cbbbb63/logTestSessionMigration1750115098
    test_log_scope.go:82: use -show-logs to present logs inline
    session_migration_test.go:239: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/17b9b212a9f8f930df1697330cbbbb63/logTestSessionMigration1750115098
--- FAIL: TestSessionMigration (8.36s)
=== RUN   TestSessionMigration/max_size
    test_server_shim.go:154: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    session_migration_test.go:92: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/16093/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/sql_test_/sql_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/testdata/session_migration/max_size:1:
          change_session_size 2KB
          ----
    session_migration_test.go:92: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/16093/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/sql_test_/sql_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/testdata/session_migration/max_size:4:
          let $large_string
          SELECT repeat('s', 5000)
          ----
    session_migration_test.go:92: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/16093/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/sql_test_/sql_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/testdata/session_migration/max_size:8:
          exec
          PREPARE q AS SELECT '$large_string'
          ----
    session_migration_test.go:233: expected error
    datadriven.go:343: 
        fatal during /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/16093/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/sql_test_/sql_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/testdata/session_migration/max_size:12:
        SELECT encode(crdb_internal.serialize_session(), 'hex')
--- FAIL: TestSessionMigration/max_size (1.08s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSessionMigration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-56286

--- a/pkg/sql/session_migration_test.go
+++ b/pkg/sql/session_migration_test.go
@@ -14,6 +14,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/datapathutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
@@ -21,6 +22,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/datadriven"
+	"github.com/cockroachdb/errors"
 	"github.com/jackc/pgx/v5"
 	"github.com/stretchr/testify/require"
 )
@@ -206,6 +208,21 @@ WHERE dump.variable IS NULL OR dump2.variable IS NULL OR dump.variable != dump2.
 				if err != nil {
 					return err.Error()
 				}
+				// Block until the application layer observes the change.
+				testutils.SucceedsSoon(t, func() error {
+					row := dbConn.QueryRow(ctx, `SHOW CLUSTER SETTING sql.session_transfer.max_session_size;`)
+					var sessionSize string
+					err := row.Scan(&sessionSize)
+					if err != nil {
+						return err
+					}
+					// Remove spaces to match format from the test file.
+					sessionSize = strings.ReplaceAll(sessionSize, " ", "")
+					if sessionSize != d.CmdArgs[0].Key {
+						return errors.Newf("expected session size to be %s, got %s", d.CmdArgs[0].Key, sessionSize)
+					}
+					return nil
+				})
 				return ""
 
 			case "error":

--- a/pkg/sql/testdata/session_migration/max_size
+++ b/pkg/sql/testdata/session_migration/max_size
@@ -1,4 +1,4 @@
-change_session_size 2KB
+change_session_size 2.0KiB
 ----
 
 let $large_string

